File: scripts/enum/sts.js
'use strict';

const { STSClient, GetCallerIdentityCommand } = require('@aws-sdk/client-sts');
const {
  OrganizationsClient,
  DescribeOrganizationCommand,
  ListPoliciesCommand,
  DescribePolicyCommand,
  ListTargetsForPolicyCommand,
} = require('@aws-sdk/client-organizations');

const { withRetry, paginate, createLogger } = require('../lib');
const { baseEnum } = require('../lib/base-enum');

// --- Principal type detection ---

function derivePrincipalType(arn) {
  if (!arn) return 'unknown';
  if (arn.includes(':assumed-role/')) return 'assumed-role';
  if (arn.includes(':role/')) return 'role';
  if (arn.includes(':user/')) return 'user';
  if (arn.includes(':federated-user/')) return 'federated-user';
  if (arn.includes(':root')) return 'root';
  return 'unknown';
}

// --- Run (DI-injectable) ---

async function run(opts = {}) {
  const { runDir } = opts;
  const stsClient = opts.clients && opts.clients.sts ? opts.clients.sts : new STSClient({});
  const orgsClient = opts.clients && opts.clients.organizations ? opts.clients.organizations : new OrganizationsClient({});

  const logger = opts.logger || createLogger(runDir, 'sts');
  const findings = [];
  let status = 'complete';

  // --- Step 1: GetCallerIdentity (fatal on failure) ---
  let identity;
  try {
    logger.log('api_call', 'GetCallerIdentity', { service: 'sts' });
    identity = await withRetry(() => stsClient.send(new GetCallerIdentityCommand({})));
  } catch (err) {
    logger.log('error', 'GetCallerIdentity', { error: err.message });
    await logger.flush();
    console.error(`FATAL: GetCallerIdentity failed: ${err.message}`);
    throw err;
  }

  const accountId = identity.Account;
  const callerArn = identity.Arn;
  const userId = identity.UserId;

  findings.push({
    resource_type: 'sts_identity',
    resource_id: 'caller',
    arn: callerArn,
    region: 'global',
    account_id: accountId,
    user_id: userId,
    principal_type: derivePrincipalType(callerArn),
    findings: [],
  });

  logger.log('info', 'CallerIdentity resolved', { account_id: accountId, arn: callerArn });

  // --- Step 2: Organizations discovery (best-effort) ---
  let orgAccessible = false;

  try {
    logger.log('api_call', 'DescribeOrganization', { service: 'organizations' });
    const orgResp = await withRetry(() => orgsClient.send(new DescribeOrganizationCommand({})));
    const org = orgResp.Organization;

    orgAccessible = true;
    findings.push({
      resource_type: 'sts_organization',
      resource_id: org.Id,
      arn: org.Arn,
      region: 'global',
      master_account_id: org.MasterAccountId,
      available_policy_types: (org.AvailablePolicyTypes || [])
        .filter((pt) => pt.Status === 'ENABLED')
        .map((pt) => pt.Type),
      findings: [],
    });

    logger.log('info', 'Organization discovered', { org_id: org.Id });
  } catch (err) {
    const code = err.name || err.Code || '';
    if (
      code === 'AccessDeniedException' ||
      code === 'AWSOrganizationsNotInUseException' ||
      code === 'AccessDenied'
    ) {
      logger.log('info', 'Organizations not accessible', { error: code });
      status = 'partial';
    } else {
      logger.log('error', 'DescribeOrganization unexpected error', { error: err.message });
      status = 'partial';
    }
  }

  // --- Step 3: SCP enumeration (only if org is accessible) ---
  if (orgAccessible) {
    try {
      logger.log('api_call', 'ListPolicies', { service: 'organizations', filter: 'SERVICE_CONTROL_POLICY' });
      const policies = await paginate(orgsClient, ListPoliciesCommand, 'Policies', {
        params: { Filter: 'SERVICE_CONTROL_POLICY' },
      });

      for (const policySummary of policies) {
        try {
          // Get full policy document
          logger.log('api_call', 'DescribePolicy', { service: 'organizations', policy_id: policySummary.Id });
          const policyResp = await withRetry(() =>
            orgsClient.send(new DescribePolicyCommand({ PolicyId: policySummary.Id }))
          );
          const policy = policyResp.Policy;

          // Get targets for this policy
          logger.log('api_call', 'ListTargetsForPolicy', { service: 'organizations', policy_id: policySummary.Id });
          const targets = await paginate(orgsClient, ListTargetsForPolicyCommand, 'Targets', {
            params: { PolicyId: policySummary.Id },
          });

          // Parse policy document JSON
          let policyDocument = {};
          try {
            policyDocument = JSON.parse(policy.PolicySummary ? policy.Content : policy.Content);
          } catch (parseErr) {
            policyDocument = { raw: policy.Content };
          }

          findings.push({
            resource_type: 'sts_scp',
            resource_id: policySummary.Id,
            arn: policySummary.Arn,
            region: 'global',
            name: policySummary.Name,
            description: policySummary.Description || '',
            policy_document: policyDocument,
            targets: targets.map((t) => ({
              target_id: t.TargetId,
              type: t.Type,
              name: t.Name,
            })),
            findings: [],
          });
        } catch (scpErr) {
          logger.log('error', 'SCP enumeration failed for policy', {
            policy_id: policySummary.Id,
            error: scpErr.message,
          });
          status = 'partial';
        }
      }
    } catch (err) {
      logger.log('error', 'ListPolicies failed', { error: err.message });
      status = 'partial';
    }
  }

  logger.log('info', 'STS_Enumeration_Complete', { findings_count: findings.length, status });
  await logger.flush();

  return { findings, status };
}

if (require.main === module) {
  baseEnum({ module: 'sts', run, global: true });
}
module.exports = { run };

File: enumerators/sts.py
from __future__ import annotations

import json
from typing import Any

from scope_core.coverage import CoverageTracker
from scope_core.models import ModuleEnvelope


def derive_principal_type(arn: str | None) -> str:
    if not arn:
        return "unknown"
    if ":assumed-role/" in arn:
        return "assumed-role"
    if ":role/" in arn:
        return "role"
    if ":user/" in arn:
        return "user"
    if ":federated-user/" in arn:
        return "federated-user"
    if arn.endswith(":root"):
        return "root"
    return "unknown"


def _client_error_code(err: Exception) -> str:
    response = getattr(err, "response", None)
    if isinstance(response, dict):
        error = response.get("Error")
        if isinstance(error, dict) and error.get("Code"):
            return str(error["Code"])
    return err.__class__.__name__ or "Error"


def _parse_policy_document(content: str | None) -> dict[str, Any]:
    if not content:
        return {}
    try:
        parsed = json.loads(content)
    except json.JSONDecodeError:
        return {"raw": content}
    if isinstance(parsed, dict):
        return parsed
    return {"raw": parsed}


def run(factory: Any, region: str) -> ModuleEnvelope:
    tracker = CoverageTracker(primary=["get_caller_identity"], required=["scp_policy", "scp_policy_detail", "scp_targets"])
    sts = factory.client("sts")
    findings: list[dict[str, Any]] = []

    try:
        identity = sts.get_caller_identity()
        tracker.record_ok("get_caller_identity")
    except Exception as err:
        tracker.record_module_failure("get_caller_identity", operation="sts.GetCallerIdentity", err=err)
        coverage, errors = tracker.to_envelope_fields()
        return ModuleEnvelope(
            module="sts",
            account_id="unknown",
            region=region,
            status=tracker.derive_status(),
            resources=[],
            coverage=coverage,
            errors=errors,
        )

    account_id = str(identity["Account"])
    caller_arn = identity.get("Arn")
    identity_finding: dict[str, Any] = {
        "resource_type": "sts_identity",
        "resource_id": "caller",
        "arn": caller_arn,
        "region": "global",
        "account_id": account_id,
        "user_id": identity.get("UserId"),
        "principal_type": derive_principal_type(caller_arn),
        # DEPRECATED: keep for one release while Organizations state moves into coverage[]; remove in v1.16.
        "org_accessible": False,
        "org_status": "unknown",
        "org_error_code": None,
        "findings": [],
    }
    findings.append(identity_finding)

    org_accessible = False
    organizations = None
    try:
        organizations = factory.client("organizations")
        org_response = organizations.describe_organization()
        tracker.record_ok("organizations")
        org = org_response.get("Organization", {})
        org_accessible = True
        identity_finding["org_accessible"] = True
        identity_finding["org_status"] = "accessible"
        findings.append(
            {
                "resource_type": "sts_organization",
                "resource_id": org.get("Id"),
                "arn": org.get("Arn"),
                "region": "global",
                "master_account_id": org.get("MasterAccountId"),
                "available_policy_types": [
                    policy_type.get("Type")
                    for policy_type in org.get("AvailablePolicyTypes", [])
                    if policy_type.get("Status") == "ENABLED"
                ],
                "findings": [],
            }
        )
    except Exception as err:
        code = _client_error_code(err)
        identity_finding["org_error_code"] = code
        if code == "AWSOrganizationsNotInUseException":
            identity_finding["org_status"] = "not_in_use"
            tracker.record_skipped("organizations", reason=code)
        elif code in {"AccessDeniedException", "AccessDenied"}:
            identity_finding["org_status"] = "not_accessible"
            tracker.record_skipped("organizations", reason=code)
        else:
            identity_finding["org_status"] = "error"
            tracker.record_resource_failure(
                "organizations",
                resource=account_id,
                operation="organizations.DescribeOrganization",
                err=err,
            )

    if org_accessible:
        try:
            policies = factory.paginate(
                organizations,
                "list_policies",
                "Policies",
                Filter="SERVICE_CONTROL_POLICY",
            )
            tracker.record_ok("scp_policy")
        except Exception as err:
            tracker.record_module_failure(
                "scp_policy",
                operation="organizations.ListPolicies",
                err=err,
            )
            policies = []

        for policy_summary in policies:
            policy_id = policy_summary.get("Id")
            try:
                policy_response = organizations.describe_policy(PolicyId=policy_id)
                policy = policy_response.get("Policy", {})
                tracker.record_ok("scp_policy_detail", resource=policy_id)
            except Exception as err:
                tracker.record_resource_failure(
                    "scp_policy_detail",
                    resource=policy_id,
                    operation="organizations.DescribePolicy",
                    err=err,
                )
                continue

            try:
                targets = factory.paginate(
                    organizations,
                    "list_targets_for_policy",
                    "Targets",
                    PolicyId=policy_id,
                )
                tracker.record_ok("scp_targets", resource=policy_id)
            except Exception as err:
                tracker.record_resource_failure(
                    "scp_targets",
                    resource=policy_id,
                    operation="organizations.ListTargetsForPolicy",
                    err=err,
                )
                targets = []

            findings.append(
                {
                    "resource_type": "sts_scp",
                    "resource_id": policy_id,
                    "arn": policy_summary.get("Arn"),
                    "region": "global",
                    "name": policy_summary.get("Name"),
                    "description": policy_summary.get("Description") or "",
                    "policy_document": _parse_policy_document(policy.get("Content")),
                    "targets": [
                        {
                            "target_id": target.get("TargetId"),
                            "type": target.get("Type"),
                            "name": target.get("Name"),
                        }
                        for target in targets
                    ],
                    "findings": [],
                }
            )

    coverage, errors = tracker.to_envelope_fields()
    return ModuleEnvelope(
        module="sts",
        account_id=account_id,
        region=region,
        status=tracker.derive_status(),
        resources=findings,
        coverage=coverage,
        errors=errors,
    )
